src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,185 @@
+                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
+                {
+                    context.ConnectionStream = isAsync
+                        ? await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false)
+                        : HandleTcpRequest(context, isAsync, ct).Result;
+                }
+            }
+            catch (Exception e)
+            {
+                context.Error = e;
+            }
+
+
+            if (NextHandler is not null)
+            {
+                if (isAsync)
+                {
+                    await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+                }
+                else
+                {
+                    NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async ValueTask<Stream> HandleTcpRequest(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(context.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(context.DataSource.ServerName);
+            if (ipAddresses.Length == 0)
+            {
+                throw new Exception("Hostname did not resolve");
+            }
+
+            // If there is an IP version preference, apply it
+            switch (context.IpAddressPreference)
+            {
+                case SqlConnectionIPAddressPreference.IPv4First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV4);
+                    break;
+
+                case SqlConnectionIPAddressPreference.IPv6First:
+                    Array.Sort(ipAddresses, IpAddressVersionSorter.InstanceV6);
+                    break;
+
+                case SqlConnectionIPAddressPreference.UsePlatformDefault:
+                default:
+                    // Not sorting necessary
+                    break;
+            }
+
+            // Attempt to connect to one of the matching IP addresses
+            // @TODO: Handle opening in parallel
+            Socket socket = null;
+            var socketOpenExceptions = new List<Exception>();
+
+            var ipEndpoint = new IPEndPoint(IPAddress.None, context.DataSource.Port); // Allocate once
+            foreach (var ipAddress in ipAddresses)
+            {
+                ipEndpoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocketAsync(ipEndpoint, ct).ConfigureAwait(false);
```

Comment:
What about when isAsync is false?

Reply:
WE definitely need a `if/else` here on `isAsync` flag. For the else, you could choose to implement the socket connect for sync, in this PR or throw an exception in this PR, and then create another PR for sync socket open.

Reply:
Ok, addressed by having separate flows for opening socket synchronously and asynchronously.